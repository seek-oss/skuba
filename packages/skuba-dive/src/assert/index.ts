import assert from 'assert';

/**
 * Assert that a value is either `null` or `undefined`.
 */
export function nullish<T>(
  value: T | null | undefined,
): asserts value is null | undefined {
  assert.strictEqual(typeof value === 'undefined' || value === null, true);
}

/**
 * Assert that a value is neither `null` nor `undefined`.
 */
export function notNullish<T>(value: T | null | undefined): asserts value is T {
  assert.notStrictEqual(value, null);
  assert.notStrictEqual(typeof value, 'undefined');
}

/**
 * Assert that a value is an object.
 */
export function object(
  value: unknown,
): asserts value is Record<PropertyKey, unknown> {
  assert.strictEqual(typeof value, 'object');
  assert.notStrictEqual(value, null);
}

/**
 * Assert that a value is an object with the given numeric property.
 */
export function numberProp<P extends PropertyKey>(
  value: unknown,
  prop: P,
): asserts value is Record<P, string> {
  object(value);
  assert.strictEqual(typeof value[prop], 'number');
}

/**
 * Assert that a value is an object with the given string property.
 */
export function stringProp<P extends PropertyKey>(
  value: unknown,
  prop: P,
): asserts value is Record<P, string> {
  object(value);
  assert.strictEqual(typeof value[prop], 'string');
}
