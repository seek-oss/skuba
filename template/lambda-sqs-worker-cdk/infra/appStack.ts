import { LambdaDeployment } from '@seek/aws-codedeploy-infra';
import {
  Duration,
  Stack,
  type StackProps,
  aws_iam,
  aws_kms,
  aws_lambda,
  aws_lambda_event_sources,
  aws_lambda_nodejs,
  aws_sns,
  aws_sns_subscriptions,
  aws_sqs,
} from 'aws-cdk-lib';
import type { Construct } from 'constructs';

import { config } from './config';

export class AppStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const accountPrincipal = new aws_iam.AccountPrincipal(this.account);

    const kmsKey = new aws_kms.Key(this, 'kms-key', {
      description: '<%- serviceName %>',
      enableKeyRotation: true,
      admins: [accountPrincipal],
      alias: 'seek/self/<%- serviceName %>',
    });

    kmsKey.grantEncrypt(accountPrincipal);

    const deadLetterQueue = new aws_sqs.Queue(
      this,
      'worker-queue-dead-letters',
      {
        queueName: '<%- serviceName %>-dead-letters',
        encryptionMasterKey: kmsKey,
      },
    );

    const queue = new aws_sqs.Queue(this, 'worker-queue', {
      queueName: '<%- serviceName %>',
      deadLetterQueue: {
        maxReceiveCount: 3,
        queue: deadLetterQueue,
      },
      encryptionMasterKey: kmsKey,
    });

    const topic = aws_sns.Topic.fromTopicArn(
      this,
      'source-topic',
      config.sourceSnsTopicArn,
    );

    topic.addSubscription(new aws_sns_subscriptions.SqsSubscription(queue));

    const architecture = '<%- lambdaCdkArchitecture %>';

    const defaultWorkerConfig: aws_lambda_nodejs.NodejsFunctionProps = {
      architecture: aws_lambda.Architecture[architecture],
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      environmentEncryption: kmsKey,
      // aws-sdk-v3 sets this to true by default, so it is not necessary to set the environment variable
      // https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-reusing-connections.html
      awsSdkConnectionReuse: false,
    };

    const defaultWorkerBundlingConfig: aws_lambda_nodejs.BundlingOptions = {
      sourceMap: true,
      target: 'node20',
      // aws-sdk-v3 is set as an external module by default, but we want it to be bundled with the function
      externalModules: [],
    };

    const defaultWorkerEnvironment: Record<string, string> = {
      NODE_ENV: 'production',
      // https://nodejs.org/api/cli.html#cli_node_options_options
      NODE_OPTIONS: '--enable-source-maps',
    };

    const worker = new aws_lambda_nodejs.NodejsFunction(this, 'worker', {
      ...defaultWorkerConfig,
      entry: './src/app.ts',
      timeout: Duration.seconds(30),
      bundling: defaultWorkerBundlingConfig,
      functionName: '<%- serviceName %>',
      environment: {
        ...defaultWorkerEnvironment,
        ...config.workerLambda.environment,
      },
      // https://github.com/aws/aws-cdk/issues/28237
      // This forces the lambda to be updated on every deployment
      // If you do not wish to use hotswap, you can remove the new Date().toISOString() from the description
      description: `Updated at ${new Date().toISOString()}`,
      reservedConcurrentExecutions: config.workerLambda.reservedConcurrency,
    });

    const workerDeployment = new LambdaDeployment(this, 'workerDeployment', {
      lambdaFunction: worker,
    });

    workerDeployment.alias.addEventSource(
      new aws_lambda_event_sources.SqsEventSource(queue),
    );
  }
}
