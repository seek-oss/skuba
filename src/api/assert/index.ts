import assert from 'assert';

/**
 * Check that a value is neither `null` nor `undefined`.
 */
export function notNullish<T>(value: T | null | undefined): asserts value is T {
  assert.notStrictEqual(value, null);
  assert.notStrictEqual(typeof value, 'undefined');
}
