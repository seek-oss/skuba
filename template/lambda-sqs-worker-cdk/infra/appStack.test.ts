import { SynthUtils } from '@aws-cdk/assert';
import { App } from 'aws-cdk-lib';

import cdkJson from '../cdk.json';

import { AppStack } from './appStack';

const contexts = [
  {
    stage: 'dev',
    ...cdkJson.context,
  },

  {
    stage: 'prod',
    ...cdkJson.context,
  },
];

it.each(contexts)(
  'returns expected CloudFormation stack for $stage',
  (context) => {
    const app = new App({ context });

    const stack = new AppStack(app, 'appStack');

    const json = JSON.stringify(SynthUtils.toCloudFormation(stack)).replace(
      /"S3Key":"([0-9a-f]+)\.zip"/g,
      (_, hash) => `"S3Key":"${'x'.repeat(hash.length)}.zip"`,
    );

    expect(JSON.parse(json)).toMatchSnapshot();
  },
);
