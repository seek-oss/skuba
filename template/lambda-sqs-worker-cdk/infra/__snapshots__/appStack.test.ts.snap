// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns expected CloudFormation stack for dev 1`] = `
{
  "Resources": {
    "kmskey49FBC3B3": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Description": "serviceName",
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:GenerateDataKey",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "kmskeyAlias39245779": {
      "Properties": {
        "AliasName": "alias/seek/self/serviceName",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "topic69831491": {
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "TopicName": "serviceName",
      },
      "Type": "AWS::SNS::Topic",
    },
    "worker28EA3E30": {
      "DependsOn": [
        "workerServiceRoleDefaultPolicyBA498553",
        "workerServiceRole2130CC7F",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.zip",
        },
        "Environment": {
          "Variables": {
            "NODE_ENV": "production",
            "NODE_OPTIONS": "--enable-source-maps",
            "SOMETHING": "dev",
          },
        },
        "FunctionName": "serviceName",
        "Handler": "app.handler",
        "KmsKeyArn": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "Role": {
          "Fn::GetAtt": [
            "workerServiceRole2130CC7F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "workerServiceRole2130CC7F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "workerServiceRoleDefaultPolicyBA498553": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "workerqueueA05CE5C6",
                  "Arn",
                ],
              },
            },
            {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "kmskey49FBC3B3",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "workerServiceRoleDefaultPolicyBA498553",
        "Roles": [
          {
            "Ref": "workerServiceRole2130CC7F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "workerSqsEventSourceappStackworkerqueue8281B9F47B9F582B": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "workerqueueA05CE5C6",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "worker28EA3E30",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "workerqueueA05CE5C6": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "QueueName": "serviceName",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "workerqueuedlq42262778",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "workerqueuePolicy97054CB4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "topic69831491",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "workerqueueA05CE5C6",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "workerqueueA05CE5C6",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "workerqueueappStacktopic0CA45134AFB31FF4": {
      "DependsOn": [
        "workerqueuePolicy97054CB4",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "workerqueueA05CE5C6",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": {
          "Ref": "topic69831491",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "workerqueuedlq42262778": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "QueueName": "serviceName-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`returns expected CloudFormation stack for prod 1`] = `
{
  "Resources": {
    "kmskey49FBC3B3": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Description": "serviceName",
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:GenerateDataKey",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "kmskeyAlias39245779": {
      "Properties": {
        "AliasName": "alias/seek/self/serviceName",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "topic69831491": {
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "TopicName": "serviceName",
      },
      "Type": "AWS::SNS::Topic",
    },
    "worker28EA3E30": {
      "DependsOn": [
        "workerServiceRoleDefaultPolicyBA498553",
        "workerServiceRole2130CC7F",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.zip",
        },
        "Environment": {
          "Variables": {
            "NODE_ENV": "production",
            "NODE_OPTIONS": "--enable-source-maps",
            "SOMETHING": "prod",
          },
        },
        "FunctionName": "serviceName",
        "Handler": "app.handler",
        "KmsKeyArn": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "Role": {
          "Fn::GetAtt": [
            "workerServiceRole2130CC7F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "workerServiceRole2130CC7F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "workerServiceRoleDefaultPolicyBA498553": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "workerqueueA05CE5C6",
                  "Arn",
                ],
              },
            },
            {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "kmskey49FBC3B3",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "workerServiceRoleDefaultPolicyBA498553",
        "Roles": [
          {
            "Ref": "workerServiceRole2130CC7F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "workerSqsEventSourceappStackworkerqueue8281B9F47B9F582B": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "workerqueueA05CE5C6",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "worker28EA3E30",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "workerqueueA05CE5C6": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "QueueName": "serviceName",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "workerqueuedlq42262778",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "workerqueuePolicy97054CB4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "topic69831491",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "workerqueueA05CE5C6",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "workerqueueA05CE5C6",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "workerqueueappStacktopic0CA45134AFB31FF4": {
      "DependsOn": [
        "workerqueuePolicy97054CB4",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "workerqueueA05CE5C6",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": {
          "Ref": "topic69831491",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "workerqueuedlq42262778": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "kmskey49FBC3B3",
            "Arn",
          ],
        },
        "QueueName": "serviceName-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
